[quant] conv_transpose2d_prepack/conv_transpose2d_unpack

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43199 [quant] conv_transpose graph patterns
* #40371 [quant] ConvTranspose1d / ConvTranspose2d
* #40370 [quant] conv_transpose1d / conv_transpose2d
* #40360 [quant] conv_transpose1d_prepack / conv_transpose1d_unpack
* **#40351 [quant] conv_transpose2d_prepack/conv_transpose2d_unpack**
* #39714 [quant] Prep for conv_transpose packing

Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

Commit:
[quant] conv_transpose2d_prepack/conv_transpose2d_unpack

[ghstack-poisoned]

z-a-f commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -226,7 +231,14 @@ c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> PackedConvWeightsQnnp<
       bias_fp32.sizes(),
-      " instead");
+      " instead. "
+      "(weight dimensions: ",
+      weight.sizes(), " , transpose: ",
+      (transpose ? "True)." : "False)."),
+      "\nDEBUG INFO: ",
+      bias_fp32.ndimension(), ", ",
+      bias_fp32.size(0), ", ", out_ch
```
Remove debug information

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"




[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"




[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"




[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"




[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -378,15 +413,21 @@ class QConv1dPackWeightInt8 final {
+  // Conv
   m.impl("conv2d_prepack", TORCH_FN(QConvPackWeightInt8<2>::run_conv));
+  // ConvTranspose
+  m.impl("conv_transpose2d_prepack", TORCH_FN(QConvPackWeightInt8<2>::run_deconv));
```
just curious, what's the reason to add this twice (also on line 423)?

z-a-f replied:
If I understand it correctly, the `_` prefix is still used, and will be removed once the ops are migrated to mobile. However, before this is done, it should be registered for both (I think): https://github.com/pytorch/pytorch/issues/36510

z-a-f replied:
I wonder if we still need to do it? @jerryzh168

vkuzo replied:
maybe let's follow up on if we need this?

z-a-f replied:
I am fine either way -- the only concern is that if removed it will break the consistency with respect to the `conv_Xd` methods.

vkuzo commented on aten/src/ATen/native/quantized/cpu/qconv_unpack.cpp:
```diff
@@ -204,6 +204,15 @@ class QConvPadding final {
+class QConvOutputPadding final {
+ public:
+  static torch::List<int64_t> run(
+      const c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>>& packed_weight) {
```
nit: `packed_params`, since it's actually not the weight that's being passed in?

z-a-f replied:
Although we call it `ConvPackedParams`, I think on python level it is still called `packed_weight`. I can change it to the `params` -- I am just concerned if this would be equally confusing?

vkuzo reviewed (commented):
looks reasonable, ideally we can review again after serialization is clear

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qconv_unpack.cpp:
```diff
@@ -230,14 +248,30 @@ TORCH_LIBRARY_IMPL(quantized, CatchAll, m) {
+  m.impl("conv3d_transpose", TORCH_FN(QConvTranspose<3>::run));
+
+  // ConvTranspose is the same, however, we want to have different name.
+  m.impl("conv_transpose2d_unpack", TORCH_FN(QConvUnpackWeightsInt8<2>::run));
+
+  m.impl("conv_transpose2d_stride", TORCH_FN(QConvStride<2>::run));
+  m.impl("conv_transpose2d_padding", TORCH_FN(QConvPadding<2>::run));
+  m.impl("conv_transpose2d_output_padding", TORCH_FN(QConvOutputPadding<2>::run));
+  m.impl("conv_transpose2d_dilation", TORCH_FN(QConvDilation<2>::run));
+  m.impl("conv_transpose2d_groups", TORCH_FN(QConvGroups<2>::run));
+  m.impl("conv_transpose2d_transpose", TORCH_FN(QConvTranspose<2>::run));
```
why do we need different names?

z-a-f replied:
You mean in addition to the `conv2d_*`? If so, this is just for consistency. They call the same functions, but because we have `conv2d` and `conv_transpose2d`, I thought it would be confusing to have `conv2d_*`, but not the `conv_transpose2d*`. LMK if I should change it

vkuzo replied:
maybe we can avoid the duplicate methods and add a comment?

z-a-f replied:
Same as my previous comment -- I am fine removing it, the only convern is the symmetry with respect to the conv_Xd methods.

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp:
```diff
@@ -284,7 +292,19 @@ class QConvPackWeightInt8 final {
+      torch::List<int64_t> dilation,
+      int64_t groups) {
+    return _run(weight, bias, stride, padding, output_padding, dilation, groups,
+                /*transpose=*/true);
```
optional nit, would be good to avoid `/*....*/`, this removes the ability to easily comment out blocks of code for debugging

Commit:
Update on "[quant] conv_transpose2d_prepack/conv_transpose2d_unpack"


Differential Revision: [D22158983](https://our.internmc.facebook.com/intern/diff/D22158983)

[ghstack-poisoned]
